src/net/libwebsocketsIO.cpp

```diff
@@ -297,9 +297,19 @@ int LibwebsocketsClient::wsCallback(struct lws *wsi, enum lws_callback_reasons r
                 WEBSOCKETS_LOG_DEBUG("Forced disconnect completed");
                 return -1;
             }
-            
-            WEBSOCKETS_LOG_DEBUG("Graceful disconnect completed");
             client->disconnecting = false;
+            if (client->wsIsConnected())
+            {
+                WEBSOCKETS_LOG_DEBUG("Disconnect done by server");
+                struct lws *dwsi = client->wsi;
+                client->wsi = NULL;
+                lws_set_wsi_user(dwsi, NULL);
+            }
+            else
+            {
```

Comment:
This code doesn't seem reachable because `wsIsConnected()` checks `client->wsi` and I think that it's only `NULL` when the `user` pointer is also `NULL` and there is an `if` above that checks if the `user` pointer is `NULL`,

I think that it would be more appropriate to use one log or the other depending on the value of `client->disconnecting` (it should be `true` before it's set to `false` above for graceful disconnects only).

Also, I would unconditionally set `client->wsi` and the `user` pointer to `NULL`.

Somethink like:
```
if (client->disconnecting)
{
    WEBSOCKETS_LOG_DEBUG("Graceful disconnect completed");
    client->disconnecting = false;
}
else
{
    WEBSOCKETS_LOG_DEBUG("Disconnect done by server");
}

struct lws *dwsi = client->wsi;
client->wsi = NULL;
lws_set_wsi_user(dwsi, NULL);
```

Of course, that is only valid of `client->wsi` can't be `NULL` here.
A safer approach could be to use the `client->disconnecting` to select the log, but still check if `client->wsi` is `NULL` before using it.

Also, checking the code of libwebsockets, it seems safe to not use `dwsi`, because it seems that `lws_set_wsi_user` can't trigger a callback, but anyway it's not bad to keep it to be protected about possible changes in that function.

So basically, I would just change the selection of the log in this snippet. The rest of proposed changes shouldn't change the behavior of this code at all.

Reply:
I agree with you. I'll change the way we log the message to be based on the `disconnecting`, and keep the `if (client->wsIsConnected())` to set the `client->wsi = NULL`